compiler/qsc_hir/src/ty.rs

```diff
@@ -61,28 +61,68 @@ impl Ty {
                     .map(|item| item.with_package(package))
                     .collect(),
             ),
-            Ty::Udt(res) => Ty::Udt(res.with_package(package)),
+            Ty::Udt(name, res) => Ty::Udt(name.clone(), res.with_package(package)),
+        }
+    }
+
+    pub fn display(&self) -> String {
+        match self {
+            Ty::Array(item) => {
+                format!("{}[]", item.display())
+            }
+            Ty::Arrow(arrow) => {
+                let arrow_symbol = match arrow.kind {
+                    CallableKind::Function => "->",
+                    CallableKind::Operation => "=>",
+                };
+                let functors = if arrow.functors == FunctorSet::Value(FunctorSetValue::Empty) {
+                    String::new()
+                } else {
+                    format!(" is {}", arrow.functors)
+                };
+                format!(
+                    "({} {arrow_symbol} {}{functors})",
+                    arrow.input.display(),
+                    arrow.output.display()
+                )
+            }
+            Ty::Infer(_) | Ty::Err => "?".to_string(),
```

Comment:
One minor advantage of including the inference integer in the error is it lets the user know what types are distinct even if they are unknown. I don't know if that's worth the effort of figuring out how to display them consistently, but it is some information that user previous could get from error messages that will be lost with this pattern.

Reply:
That's true, and I think it is a balancing act of trying to figure out what is most helpful to the user. I lean slightly on the side of thinking that that information is not usually very relevant to a user and the single `?` is prettier. I could be convinced that that information is very useful to a user though. I'd also like to get @minestarks to weigh in on this change, whether or not to go forward with it.

Reply:
Well, if my opinion is the tiebreaker, I still cast my vote with the `?` display. While the numbers are helpful to people who "know what they're doing" (i.e. us), I don't think the average programmer knows enough about type inference to understand what the numbers are for, and the confusion they create negates any clarification they bring. Again, my opinion and not a very strong one.

Reply:
Oh and other mature languages (TypeScript, Rust) which do pretty sophisticated type inference also just show "unknown" without exposing any other inference info. This tells me that there just isn't much user demand for this type of thing.

Reply:
That makes sense to me. I don't feel strongly about it either, so just "?" seems good!

Reply:
Okay, I'm going forward with the single `?` then. This can easily be changed in the future if the demand for more specific information surfaces.